Docs: API Doc hackathon project (#209)
grafana/writers-toolkit: docs/sources/style-guide/write-for-developers/index.md

---
title: Write for developers

> description: Guidelines on creating documentation for software developers and engineers

Reviewer: ```suggestion
description: Write documentation for software developers and engineers.
```

aliases:
  - /docs/writers-toolkit/latest/style-guide/writing-for-developers/
weight: 500
keywords:
  - Grafana
  - Writing for developers
  - Docs for developers
---

# Write for developers

> These guidelines provide suggestions for creating documentation for software develoers and engineers. Follow these tips to write useful API docs, code examples, and other technical material. Read these guidelines in the context of [Grafana's style guide]({{< relref "../../style-guide/" >}}) to learn how to communicate effectively with the developers who enhance and work with the code of our project. 

Reviewer: ```suggestion
The guidelines that follow provide suggestions for writing documentation for software developers and engineers. 
Follow these tips to write useful API documentation, code examples, and other technical material.
To learn how to communicate effectively with the developers who enhance and work with the code of Grafana Labs projects or products, read these guidelines in the context of the [Style guide]({{< relref "../../style-guide/" >}}).
```
Rationale:
* Added semantic line breaks for easier editing later on.
* Changed "creating" to "writing" to use one verb instead of two for the same thing.
* Expanded "doc" to "documentation" in long-form text.
* For cross-reference, I switched the word order so that the WHAT comes before the HOW.
* The link text is now WYSIWYG.

## Developer documentation basics

> Writing technical content for software developers is similar to writing content for users or administrators of software products, so they follow the same general guidelines. However, developer-facing docs tend to be more technical, and they rely upon certain conventions that are important to recognize. They often talk about code, so it's important to know how to structure, format, and identify common problems that may arise with it in documentation. 

Reviewer: ```suggestion
Writing technical content for software developers is similar to writing content for users or administrators of software products, so the same general guidelines apply.
However, developer-facing documentation tends to be more technical, and relies on certain important conventions.
Because this type of documentation contains details about code, it's important to know how to structure, format, and identify common problems that might arise.
```

Rationale:
* Added semantic line breaks for easier editing later on.
* Reword ambiguous "they".
* In general, spell out words in long-form text; exceptions include contractions.
* Changed "upon" to "on" to remove connotation of trust, and to make it a tad less formal.
* Shortened length of sentence by using an adjective.
* Reworded (in this case made longer) ambiguous "they".
* Changed "may" to "might" to move connotations from permission to possibility.
* Shortened last sentence because the final phrase is unneeded.

> When readers of your documents are developers, you can assume that they are familiar with general programming concepts. There is no need to spell out elementary ideas. Instead, talk about those concepts and features that are specific to our products. For instance, instead of talking about the fundamentals of UI design  _in general_, talking about how Grafana's software or APIs _interpret_ those principles.

Reviewer: ```suggestion
When your readers are developers, you can assume that they are familiar with general programming concepts.
There is no need to explain elementary ideas. Instead, introduce those concepts and features that are specific to Grafana Labs products.
For example, instead of covering the fundamentals of UI design  _in general_, explain how Grafana Labs software or APIs _interpret_ those principles.
```
Rationale:
* Shortened first sentence; "readers" implies they are reading some sort of documentation.
* Replaced phrasal verb (to spell out) with a simpler one (explain).
* Replaced "talk about" with "introduce" because this is written material as opposed to spoken material. Alternatives include, "cover", "teach", and more.
* Replaced "our" with Grafana Labs because it feels less like "ours" vs. "yours", and if someone in the OSS community is writing documentation, it feels a bit off. (Are we expecting any external contributions?)
* Replaced "instance" with "example" for consistency, and also because the term "instance" is overloaded otherwise; "... an instance of Prometheus..."
*  Replaced all instances of Grafana with Grafana Labs because Grafana is not the only OSS project of Grafana Labs.

## Code comments

> The foundation of strong documentation is well-written comments in code: concise, relevant, and up-to-date. 

Reviewer: ```suggestion
The foundation of strong documentation is well-written comments in code that are concise, relevant, and current. 
```

> This guide doesn't cover the do's and don'ts of writing comments, but you can refer to the [Grafana developer style guide](https://github.com/grafana/grafana/blob/main/contribute/style-guides/code-comments.md).

Reviewer: WYSIWYG link text

> For more general advice, consult a reputable [coding style guide](https://google.github.io/styleguide/) for your favorite programming language. 

Reviewer: ```suggestion
For more general advice, consult one of the reputable [Google Style Guides](https://google.github.io/styleguide/) for your favorite programming language. 
```
WYSIWYG link text

## Reference docs

> Whenever possible, API docs and other reference docs should be automatically generated from source code comments. However they happen to be created, they ought to conform to our [style conventions]({{< relref "../../style-guide/style-conventions/" >}}), with particular attention paid to properly formatting the elements of code.

Reviewer: ```suggestion
Whenever possible, automatically generate API and other reference documentation from source-code comments. However documentation is created, make sure it conforms to the [style conventions]({{< relref "../../style-guide/style-conventions/" >}}).
Pay particular attention to properly formatting the elements of code.
```

### Auto-generated documentation

The advantages of automating documentation programmatically are well-known, mainly increases in consistency and reduction of human error. But behind every line of auto-gen content is a human author, one who is responsible for following our [style guidelines]({{< relref "../../style-guide/" >}}).

> When writing docs that will be used by an auto-gen program to create publishable content, keep the following in mind:

Reviewer: ```suggestion
When writing documentation that will be used by an auto-generated program to create publishable content, keep the following things in mind:
```

> An auto-gen tool can parse syntax, but when writing docs that will be used to create automated content, it falls on your shoulders to add actionable insights. 

Reviewer: ```suggestion
An auto-generation tool can parse syntax, but when writing documentation that will be used to create automated content, it is up to you to add actionable insights. 
```
Replaced "falls on your shoulders" with a simpler phrase.

Reviewer: "... your responsibility... " sounded too negative to me, as an alternative.

- What are the caveats, edge cases, and side effects?

> - What is the bigger picture not self-evident in the code? 

Reviewer: ```suggestion
- What is the bigger picture that is not self-evident in the code? 
```

> - In short, what is everything that a developer needs to know to make use of the code?

Reviewer: ```suggestion
- In short, what is everything that a developer needs to know to use the code?
```

> Be aware that it may be difficult to integrate auto-gen content with other documentation such as relevant sections of getting started guides, tutorials, and detailed code examples. You may need to consult with a Doc Squad member to ensure that your content is properly cross-referenced.

Reviewer: ```suggestion
> **Note:** It might be difficult to integrate auto-generated content with other documentation, such as relevant sections of _Get started_ guides, tutorials, or detailed code examples. You might need to ask a Docs team member to ensure that your content is properly cross-referenced.
```

### Elements of API references

> Pay attention to properly documenting the most common elements of an API reference such as the title, parameters, return values, and so on. The following are suggestions that will help you to write more complete and consistent docs.

Reviewer: ```suggestion
Properly document the most common elements of an API reference, such as the title, parameters, return values, and so on. The following suggestions will help you to write more complete and consistent documentation:
```

An auto-generation tool can parse syntax, but when writing documentation that will be used to create automated content, it is up to you to add actionable insights. 
- What are the caveats, edge cases, and side effects?
- What is the bigger picture that is not self-evident in the code? 
- In short, what is everything that a developer needs to know to use the code?

> **Note:** It might be difficult to integrate auto-generated content with other documentation, such as relevant sections of _Get started_ guides, tutorials, or detailed code examples. You might need to ask a Docs team member to ensure that your content is properly cross-referenced.

### Elements of API references

Properly document the most common elements of an API reference, such as the title, parameters, return values, and so on. The following suggestions will help you to write more complete and consistent documentation:

| Element                                               | Description                                                           |
| ------------------------------------------------- | ----------------------------------------------------------------------- |

> | Title and description       | The name of the element and a description of it in one or two sentences. Put API names, classes, methods, etc. in `code font`. |

Reviewer: ```suggestion
| Title and description       | The name of the element and a description of it in one or two sentences. Use back ticks (``) for API names, classes, methods, and so forth, so they display in a fixed-width font. |
```

> | Title and description       | The name of the element and a description of it in one or two sentences. Put API names, classes, methods, etc. in `code font`. |

Reviewer 2: ```suggestion
| Title and description       | The name of the element and a description of it in one or two sentences. Put API names, classes, methods, and so forth in `monospace`. |
```

Reviewer 2: Note: we could say `code font`, but I've used `monospace` most places. I won't edit that nit further, based on what you decide.

Author: I would argue for going with `code font` because technically it is produced with a backtick in Markdown, and backticks are semantic markers for code, not for a monospace style.

Reviewer: I have usually called this `fixed-width` font.

Author: changed to "fixed-width font" for now.

| Syntax | The code signature that defines the element. If multiple programming languages can be used, give the syntax in each. Put the signature in `code font`. |

> | Parameters | If the element has parameters, specify their descriptions, data types, and state whether they are optional or required. Put the parameters in _italic_. |

Author: I don't know what conversations have been had, but the Docs team has specified Google as a fallback guide for cases where we haven't chiimed in, and Google specifies italic. 

<PASSWORD> is a placeholder, and Google has a whole page devoted to those at https://developers.google.com/style/placeholders. I mention them very briefly elsewhere in the doc. ALL CAPS for a placeholder should be fine, but the < and > are unnecessary.

| Return values| If the element returns a value, describe the range of possibilities and the data type. |

> | Error codes| Describe errors or exceptions and the conditions under which they occur. |

Reviewer: ```suggestion
| Error codes| Describe errors or exceptions and the conditions under which they occur. If possible, provide a way to resolve the issue. |
```

> | Comments | Describe any important information that hasn't been previously included in the title, description, syntax, parameters, or return values. For example, you may explain non-obvious context, compare it to similar elements, or provide cautionary notes about potential gotchas. |

Reviewer: ```suggestion
| Comments | Describe any important information that hasn't been previously included in the title, description, syntax, parameters, or return values. For example, you might explain non-obvious context, make a comparison to similar elements, or provide cautionary notes about potential gotchas. |
```
"it" was undefined

Additional tips:

> - Remember to write concisely. Don't say "This method does this..." or "This parameter describes that..."

Reviewer 3: >Don't say "This method does this..." or "This parameter describes that..."

I'd need to know more about why you think these constructions should be avoided.  I thought the point of developer documentation was to document the methods and parameters?  🤔

Author: Changed to "Remember to write concisely. Don't say "The AddUser method adds a user." or "This method adds a user." when "Adds a user." will do."

> - Remember to write concisely. Don't say "The AddUser method adds a user." or "This method adds a user." when "Adds a user." will do. 

Reviewer: ```suggestion
- Write concisely. Don't say "The `AddUser` method adds a user." or "This method adds a user." when "Adds a user." will do.
```
Would you consider "Add a user."? The focus shifts to what you want the reader to do as opposed to talking about what the thing does.

Author: I could live with it either way, but prefer "adds". I think "method" is still the implied subject, which makes "adds" grammatically correct. But "add" seems to turn the sentence into an imperative, which is inconsistent with the other descriptions.

> - When the names of code elements are singular, don't make them plural. Instead, add a plural noun to describe them. For example, don't change `MyEvent` to `MyEvents`; refer to the `MyEvent` objects. 

Reviewer: ```suggestion
- When the name of a code element is singular, don't pluralize it. Instead, follow it by plural noun. For example, write "`MyEvent` objects" instead of `MyEvents`.
```

> - If the element does some sort of action, start the first sentence of the description with an action verb. 

Reviewer: ```suggestion
- If the element performs an action, start the first sentence of the description with an action verb. 
```

## Code examples

> It is common for readers of documentation to skim through it to find code samples they can copy and paste and run "as is". For this reason, it is worth taking the time to ensure that whenever possible the examples you provide are production-ready.

Reviewer 2: ```suggestion
Readers of documentation typically skim through it to find code samples they can copy and paste and run "as is". Because of this, we highly recommend you ensure that whenever possible the examples you provide are production-ready.
```

> Readers of documentation typically skim through it to find code samples they can copy and paste and run "as is". Because of this, we highly recommend you ensure that whenever possible the examples you provide are production-ready.

Reviewer: ```suggestion
Readers of documentation typically skim through it to find code samples they can copy and paste and run as is.
Because of this and whenever possible, provide production-ready examples.
```

> Of course, this doesn't mean you should avoid writing code examples intended merely to illustrate a point. When your intent is to explain something, it isn't necessary for your code to be runnable so long as the developer can learn from it to do something similar on their own.

Reviewer 3: I am completely confused by line 77.

Author: Not sure why it's confusing. An example is a response from a REST API. It's provided only so the developer can compare it to the output that they get. They don't run it.

Reviewer 3: I had to stop and read this a couple of times and still wasn't quite sure what you were trying to say.  Does the following capture what you were getting at?

```suggestion
However, it isn't necessary for every code example to be runnable in production.  Some code examples are written to illustrate a point so that the developer can learn how to do something similar on their own.
```

When providing an example, give a written description. You can put it either in the body of the document or as explanatory comments within the example code.

> Remember the basic rule: your explanation of the code shouldn't describe _what_ it does, but _why_ it does what it does. (Refer to [_Docs for Developers_](https://docsfordevelopers.com/)).

Reviewer: ```suggestion
Remember the basic rule: explain _why_ your code does what it does, rather that describing _what_ it does.
For an in-depth, external resource about writing developer documentation, see [_Docs for Developers_](https://docsfordevelopers.com/).
```

> Remember the basic rule: your explanation of the code shouldn't describe _what_ it does, but _why_ it does what it does. (Refer to [_Docs for Developers_](https://docsfordevelopers.com/)).

Reviewer 3: Why are you linking to a book that must to be purchased?  If the information is not freely available on the web, I would try to find another, more accessible reference to link to.  AND link to the specific topic you have in mind.  Pointing someone to a 250 page book isn't very helpful.

Author: I don't think linking to a book -- a recent one that is the only significant book to focus directly on the topic of technical writing for developer documentation -- is inappropriate here. No one needs to buy a book if they don't want to. The point is the advice isn't being pulled out of a hat, but somewhere credible that someone can research further. Still, I can look for alternative sources on the web (but then you could easily run into the credibility problem. Why is that opinion worth following?)

Author: So, I found a blog post that says basically the same thing at https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/

The author is a developer who lives with hit wife, two cats, lots of computers, and is currently studyig Ruby. 

I would rather support this claim by reference to the leading book in the field. But if someone finds a better cite, let's use it.

Reviewer 4: I don't think the statement requires a reference. It's very general, common sense and as long as it isn't a quote, I don't think you need to cite anything. I'd add any helpful reference materials in a Learn more type section if it were me.

### Formatting code examples

Here are guidelines to follow when formatting code examples.

- Use spaces, not tabs. 
- Follow Grafana's accepted [coding style guidelines](https://github.com/grafana/grafana/blob/main/contribute/style-guides/code-comments.md).)
- Wrap lines at 80 characters.
- When omitting code, use three dots (...). Don't use the ellipsis character (…).

Introduce each code sample with a sentence or paragraph to establish its context. End the introduction with a colon if it immediately precedes the sample or a period if it doesn't. 

> ### Paths, filenames and URLs

Reviewer 2: ```suggestion
### Paths, filenames, and URLs
```

> There are many types of information that should be put in code font. Among these are filenames, paths, folders, and directories. However, don't format domain names or URLs as code if you intend the user to follow the link. 

Reviewer: ```suggestion
Many types of information belong in fixed-width font. Among these are path names, file names, directories, or folders. However, don't format domain names or URLs as code if you intend the user to follow the link.
```

> While there are times when you may want to make a URL clickable, it is better to follow our [guidelines for references]({{< relref "../../writing-guide/references/" >}}).

Reviewer: ```suggestion
Although there are times when you might want to make a URL clickable, it is better to follow the guidelines for [References]({{< relref "../../writing-guide/references/" >}}).
```

## Command lines

Use the following conventions when you include commands on the command line in technical content.

- Do not assume everyone is using Linux. Make sure instructions include enough information for Windows and Mac users to successfully complete procedures.

- Do not add `$` before commands. Make it easy for users to copy and paste commands.

  - **Right:** `sudo yum install grafana`
  - **Wrong:** `$ sudo yum install grafana`

- Include `sudo` before commands that require `sudo` to work.

> For terminal examples and Grafana configuration, use a `bash` code block:

Reviewer: ```suggestion
For terminal examples and configurations, use a `bash` code block:
```
Configurations are not just Grafana configurations.

> For terminal examples and Grafana configuration, use a `bash` code block:

Reviewer 3: It would be helpful to include the syntax for the bash code block.  Remember, community members cannot go into the source code for this page to figure the syntax out because this is a private repository.

Author: The first place I would want to see that information is in the Markdown guide. Looking there, there's a section on code comments which gives an example in JavaScript. Correct me if I'm wrong, but I think the bash and console code blocks are formed in the same fashion. So, we could add to that section with more examples or put in an explanatory note.

That said, while I'm not opposed to adding that information into this section, I do think it may divert from the primary purpose of style to go into Markdown syntax.

Reviewer 3: The markdown page works too.  I just like the ease of being able to copy/paste the syntax without having to go hunt around for it.  For reference, check out how I recently updated the Admonitions section.  
https://grafana.com/docs/writers-toolkit/style-guide/style-conventions/#admonitions

```bash
sudo yum install grafana
```

> If your command-line instructions include a combination of input and output lines, use separate code blocks for input and output, and use a `console` code block for the output.

Reviewer 3: Again, it would be helpful to provide the syntax for how to properly form the console code block in a visible way.

Author: See comment above.

Reviewer: @jdbaldry, if you have not already been asked to vet this `console` bit, can you see if this is accurate? I believe it is, but want to be sure.

```bash
cat ~/.ssh/my-ssh-key.pub
```

The output is similar to the following:

```console
ssh-rsa KEY_VALUE USERNAME
```

> For HTTP request/response, use an `http` code block:

Reviewer 3: Ibid.

Author: See comment above.

```http
GET /api/dashboards/id/1/permissions HTTP/1.1
Accept: application/json
Content-Type: application/json
Authorization: Bearer eyJrIjoiT0tTcG1pUlY2RnVKZTFVaDFsNFZXdE9ZWmNrMkZYbk
```

If an argument is optional, enclose it in square brackets.

```console
ssh-rsa KEY_VALUE USERNAME [_FILENAME_]
```

> Use descriptive words and phrases when including placeholders. Try to avoid using X or XXX. In Markdown, put a backtick followed by an asterisk in front and an asterisk followed by a backtick in front. For example, (*`A_PLACEHOLDER`*).

Reviewer: ```suggestion
Use descriptive words and phrases when including placeholders, and avoid using X or XXX.
In Markdown, in front of a placeholder, use an asterisk followed by a backtick. At the end of the placeholder, use a backtick followed by an asterisk.

 For example: *`A_PLACEHOLDER`*.
```
Rationale:
* Simplify the verbiage.
* Remove the ()s so that it simplifies the example.

> For more advice on formatting command lines, see the [Google developer style guide](https://developers.google.com/style/code-syntax).

Reviewer: ```suggestion
For more information about formatting command lines, see [Document command-line syntax](https://developers.google.com/style/code-syntax).
```

| Title and description       | The name of the element and a description of it in one or two sentences. Use back ticks (``) for API names, classes, methods, and so forth, so they display in a fixed-width font. |
| Syntax | The code signature that defines the element. If it is possible to use multiple programming languages, provide the syntax for each. Put the signature in `code font`. |
| Parameters | If the element has parameters, specify their descriptions, data types, and state whether they are optional or required. Put the parameters in _italic_. |
| Return values| If the element returns a value, describe the range of possibilities and the data type. |
| Error codes| Describe errors or exceptions and the conditions under which they occur. If possible, provide a way to resolve the issue. |
| Comments | Describe any important information that hasn't been previously included in the title, description, syntax, parameters, or return values. For example, you might explain non-obvious context, make a comparison to similar elements, or provide cautionary notes about potential gotchas. |

Additional tips:

> - Write concisely. Don't say "The `AddUser` method adds a user." or "This method adds a user." when "Adds a user." will do.

Reviewer 5: i appreciate and agree with the general point here, but the particular examples chosen will actually run afoul of Go linting standards - if there's a method called `AddUser`, it is expected that the corresponding docs begin with "AddUser", allowing "A" or "The" before it.

This seems likely to create an unnecessary conflict in a Go developer's mind. But the problem isn't the principle, just the particular example - if you're able to reword that first example to illustrate the concept without incidentally bumping into this Go convention, it would sidestep the whole problem.

Author: I have replaced with:

- Remember to write concisely. Don't say "This method adds a user." when "Adds a user." will do. However, you may say, "The AddUser method adds a user." if needed to avoid a Go linting error.

### Paths, filenames, and URLs

Many types of information belong in fixed-width font. Among these are path names, file names, directories, or folders. However, don't format domain names or URLs as code if you intend the user to follow the link.

Although there are times when you might want to make a URL clickable, it is better to follow the guidelines for [References]({{< relref "../../writing-guide/references/" >}}).

## Command lines

Use the following conventions when you include commands on the command line in technical content.

- Do not assume everyone is using Linux. Make sure instructions include enough information for Windows and Mac users to successfully complete procedures.

- Do not add `$` before commands. Make it easy for users to copy and paste commands.

  - **Right:** `sudo yum install grafana`
  - **Wrong:** `$ sudo yum install grafana`

- Include `sudo` before commands that require `sudo` to work.

For terminal examples and configurations, use a `bash` code block:

```bash
sudo yum install grafana
```

> If your command-line instructions include a combination of input and output lines, use separate code blocks for input and output, and use a `console` code block for the output.

Author: I'm finding it difficult to follow this conversation in a couple of ways. @JStickler @osg-grafana. Although I personally would prefer to add a code example to the Markdown syntax page, if we are going to add it here, could someone please supply it for me? I think isn't it just

```console
comand command command
```

@jdbaldry  seems to have confirmed that a certain syntax is correct, but I can't tell from the context what he is referring to.

Author: Note Julie raised the same issue below with bash (adding this for my own reference because I resolved the comment too soon).

Reviewer 3: 1 - To install [A thing] run the following commands:

```
```bash
command #1
```
```bash
output
```
```bash
command #2
```
```bash
output
```
```bash
command #3
```
```bash
success message
```
```

Reviewer 3: Ugh, I am failing today at providing a code sample that will show in the comments the way I want it to.  =(

Reviewer 3: But the general idea is:
1 - To install [A thing] run the following commands:
Code block with command #1 
Separate code block with expected output /success message
Code block with command #2 
Separate code block with expected output /success message
Code block with command #3 
Separate code block with expected output /success message
2 - To configure [a thing]...

Author: Hi Julie, Sam Boyer reviewed this section and I updated the PR today to accomodate his suggestions. I'm not sure I succeeded at getting the code examaples to format correctly, so you're welcome to take a look at the updated PR and re-review.

Author: Resolving for now, we can modify later if necessary.

```bash
cat ~/.ssh/my-ssh-key.pub
```

The output is similar to the following:

```console
ssh-rsa KEY_VALUE USERNAME
```

For HTTP request/response, use an `http` code block:

```http
GET /api/dashboards/id/1/permissions HTTP/1.1
Accept: application/json
Content-Type: application/json
Authorization: Bearer eyJrIjoiT0tTcG1pUlY2RnVKZTFVaDFsNFZXdE9ZWmNrMkZYbk
```

If an argument is optional, enclose it in square brackets.

```console
ssh-rsa KEY_VALUE USERNAME [_FILENAME_]

> ```

Reviewer 5: if `console` is to be used for outputs, then i'm confused about why we'd be using it here, as i read "arguments" as applying to inputs

Author: I have changed this to:

```typescript
ssh-rsa KEY_VALUE USERNAME [_FILENAME_]
```

Use descriptive words and phrases when including placeholders, and avoid using X or XXX.
In Markdown, in front of a placeholder, use an asterisk followed by a backtick. At the end of the placeholder, use a backtick followed by an asterisk.

>  For example: *`A_PLACEHOLDER`*.

Reviewer 5: i feel inclined to suggest, for this case and others, that the examples should be shown in both their literal raw input and rendered markdown form. Here, that might mean something like this instead:

---

...use a backtick followed by an asterisk. For example, prefer the following raw markdown:

```
The following text is a placeholder: *`A_PLACEHOLDER`*.
```
to produce:

> The following text is a placeholder: *`A_PLACEHOLDER`*.

---

Alternatively, a table might be used showing raw inputs to more tidily separate raw inputs and rendered markdown outputs.

Otherwise, someone reading this document in rendered markdown will have to flip between viewing markdown source and the rendered document in order to be absolutely sure of the input syntax being recommended.

https://pkg.go.dev/github.com/fluhus/godoc-tricks uses this approach to quite positive effect - it immediately organically became my go-to reference for writing Go docs with the new formatting capabilities introduced in 1.19.

Author: Perhaps we can use a table format in the future, but for now I modified the code examples using the pattern that you suggested. Please re-review.

| Title and description       | The name of the element and a description of it in one or two sentences. Use back ticks (``) for API names, classes, methods, and so forth, so they display in a fixed-width font. |
| Syntax | The code signature that defines the element. If it is possible to use multiple programming languages, provide the syntax for each. Put the signature in `code font`. |
| Parameters | If the element has parameters, specify their descriptions, data types, and state whether they are optional or required. Put the parameters in _italic_. |
| Return values| If the element returns a value, describe the range of possibilities and the data type. |
| Error codes| Describe errors or exceptions and the conditions under which they occur. If possible, provide a way to resolve the issue. |
| Comments | Describe any important information that hasn't been previously included in the title, description, syntax, parameters, or return values. For example, you might explain non-obvious context, make a comparison to similar elements, or provide cautionary notes about potential gotchas. |

Additional tips:

> - Remember to write concisely. Don't say "This method adds a user." when "Adds a user." will do. If your linter requires the description to begin with the element name, you may say "The AddUser method adds a user." to avoid an error message.

Reviewer 5: i think concision makes this tip stronger. Especially since it's a tip about being concise 😂 

```suggestion
- Remember to write concisely. Don't say "This method adds a user." when "Adds a user." will do.
```

Author: @osg-grafana @jdbaldry @sdboyer - In my last commit I fixed the sample input/output for command lines. Sam's tips were essential, but I couldn't get the exact markdown he gave me to work so I used something similar. Also, I couldn't get italic to show up as italic. Please feel free to re-review if you have a chance and make additional suggestions.

Reviewer: Hi @josmperez, nothing to add. Please merge this.

Introduce each code sample with a sentence or paragraph to establish its context. End the introduction with a colon if it immediately precedes the sample, or a period if it doesn't. 

### Paths, filenames, and URLs

Many types of information belong in fixed-width font. Among these are path names, file names, directories, or folders. However, don't format domain names or URLs as code if you intend the user to follow the link.

Although there are times when you might want to make a URL clickable, it is better to follow the guidelines for [References]({{< relref "../../writing-guide/references/" >}}).

## Command lines

Use the following conventions when you include commands on the command line in technical content.

- Do not assume everyone is using Linux. Make sure instructions include enough information for Windows and Mac users to successfully complete procedures.

- Do not add `$` before commands. Make it easy for users to copy and paste commands.

  - **Right:** `sudo yum install grafana`
  - **Wrong:** `$ sudo yum install grafana`

- Include `sudo` before commands that require `sudo` to work.

For terminal examples and configurations, use a `bash` code block. In raw markdown:

> ```bash
> sudo yum install grafana
> ```

Reviewer 5: This is unfortunately not escaped in a way that "triplebacktick+bash" shows up in the rendered output. To make it work, i think you need to add another layer of extra backticks. 

`````suggestion
````
```bash
sudo yum install grafana
```
````
`````

i got the pattern from [this stack overflow discussion](https://meta.stackexchange.com/a/155333). i'm decently sure it works because GH's suggest change feature ALSO relies on triple backticks, so i had to stack them up to five levels, the suggestion seems to come through correctly. ([this](https://gist.github.com/sdboyer/054183adf0663f5f112c672fc17897f3) is the raw markdown i wrote to produce the above suggestion)
